src/pudl/analysis/allocate_net_gen.py

```diff
@@ -645,30 +1135,501 @@ def calc_allocation_fraction(gen_pm_fuel, drop_interim_cols=True):
+                "energy_source_code",
+            ].unique()
+        )
+    )
+    msw_codes_used = list(msw_codes_in_gf | msw_codes_in_bf)
+    # join these codes into a string that will be used to replace the MSW code
+    replacement_codes = ",".join(msw_codes_used)
+
+    # if MSN and MSB codes are used, replace the existing MSW value
+    if replacement_codes != "MSW":
+        # for each type of energy source column, we want to expand any "MSW" values
+        for esc_type in ["energy_", "planned_energy_", "startup_"]:
+
+            # create a column of all unique fuels in the order in which they appear (ESC 1-6, startup fuel 1-6)
+            # this column will have each fuel code separated by a comma
+            gens["unique_esc"] = [
+                ",".join(
+                    fuel for fuel in list(dict.fromkeys(fuels)) if pd.notnull(fuel)
+                )
+                for fuels in gens.loc[
+                    :,
+                    [
+                        col
+                        for col in gens.columns
+                        if col.startswith(f"{esc_type}source_code")
+                    ],
+                ].values
+            ]
+
+            # replace any MSW codes with the codes used in bf and gf
+            gens.loc[gens["unique_esc"].str.contains("MSW"), "unique_esc"] = gens.loc[
+                gens["unique_esc"].str.contains("MSW"), "unique_esc"
+            ].str.replace("MSW", replacement_codes)
+
+            # we need to create numbered energy source code columns for each fuel
+            # first, we need to identify the maximum number of fuel codes that exist for a single generator
+            max_num_esc = (gens.unique_esc.str.count(",") + 1).max()
+
+            # create a list of numbered fuel code columns
+            esc_columns_to_add = []
+            for n in range(1, max_num_esc + 1):
+                esc_columns_to_add.append(f"{esc_type}source_code_{n}")
+
+            # drop all of the existing energy source code columns
+            gens = gens.drop(
+                columns=[
+                    col
+                    for col in gens.columns
+                    if col.startswith(f"{esc_type}source_code")
+                ]
+            )
+
+            # create new numbered energy source code columns by expanding the list of unique fuels
+            gens[esc_columns_to_add] = gens["unique_esc"].str.split(",", expand=True)
+
+            # creating columns from this list sometimes replaces NaN values with "" or None
+            gens[esc_columns_to_add] = gens[esc_columns_to_add].replace("", np.NaN)
+            gens[esc_columns_to_add] = gens[esc_columns_to_add].fillna(value=np.NaN)
+
+            # drop the intermediate column
+            gens = gens.drop(columns=["unique_esc"])
+
+    return gens
+
+
+def allocate_bf_data_to_gens(
+    bf: pd.DataFrame, gens: pd.DataFrame, bga: pd.DataFrame
+) -> pd.DataFrame:
+    """Allocates bf data to the generator level.
```

Comment:
Expand bf to boiler_fuel in docstring "Allocate boiler fuel data to the generator level."